clarity/contracts/vaults-v2/arkadiko-vaults-operations-v1-1.clar

```diff
@@ -0,0 +1,172 @@
+
+    ;; Update vault data
+    (try! (as-contract (contract-call? .arkadiko-vaults-data-v1-1 set-vault owner (contract-of token) STATUS_ACTIVE collateral debt)))
+    (try! (as-contract (contract-call? .arkadiko-vaults-sorted-v1-1 reinsert owner (contract-of token) nicr prev-owner-hint next-owner-hint)))
+
+    ;; Mint or burn USDA
+    (if (is-eq debt (get debt vault))
+      false
+      (if (> debt (get debt vault))
+        (try! (as-contract (contract-call? .arkadiko-dao mint-token .usda-token (- debt (get debt vault)) owner)))
+        (try! (as-contract (contract-call? .arkadiko-dao burn-token .usda-token (- (get debt vault) debt) owner)))
+      )
+    )
+
+    ;; Deposit or withdraw collateral
+    (if (is-eq collateral (get collateral vault))
+      false
+      (if (> collateral (get collateral vault))
+        (try! (contract-call? .arkadiko-vaults-pool-active-v1-1 deposit token owner (- collateral (get collateral vault))))
+        (try! (as-contract (contract-call? .arkadiko-vaults-pool-active-v1-1 withdraw token owner (- (get collateral vault) collateral))))
+      )
+    )
+
+    ;; Get stability fees
+    (try! (contract-call? .usda-token transfer stability-fee tx-sender .arkadiko-vaults-pool-fees-v1-1 none))
+
+    (ok true)
+  )
+)
+
+(define-public (close-vault (token <ft-trait>))
+  (let (
+    (owner tx-sender)
+    (vault (contract-call? .arkadiko-vaults-data-v1-1 get-vault owner (contract-of token)))
+    (stability-fee (unwrap-panic (get-stability-fee owner (contract-of token))))
+  )
+    (asserts! (is-eq (get status vault) STATUS_ACTIVE) (err ERR_WRONG_STATUS))
+
+    ;; Update vault data
+    (try! (as-contract (contract-call? .arkadiko-vaults-data-v1-1 set-vault owner (contract-of token) STATUS_CLOSED_BY_OWNER u0 u0)))
+    (unwrap-panic (as-contract (contract-call? .arkadiko-vaults-sorted-v1-1 remove owner (contract-of token))))
+
+    ;; Burn all debt
+    (try! (as-contract (contract-call? .arkadiko-dao burn-token .usda-token (get debt vault) owner)))
+
+    ;; Get stability fees
+    (try! (contract-call? .usda-token transfer stability-fee tx-sender .arkadiko-vaults-pool-fees-v1-1 none))
+
+    ;; Withdraw collateral
+    (try! (as-contract (contract-call? .arkadiko-vaults-pool-active-v1-1 withdraw token owner (get collateral vault))))
+
+    (ok true)
+  )
+)
+
+;; ---------------------------------------------------------
+;; Helpers
+;; ---------------------------------------------------------
+
+;; Get collateral to debt ratio
+;; Get if ratio is valid (ratio > liquidation ratio)
+(define-public (get-collateral-to-debt (oracle <oracle-trait>) (owner principal) (token principal) (collateral uint) (debt uint))
+  (let (
+    (collateral-info (unwrap! (contract-call? .arkadiko-vaults-tokens-v1-1 get-token token) (err ERR_UNKNOWN_TOKEN)))
+    (price-info (unwrap-panic (contract-call? oracle fetch-price (get token-name collateral-info))))
```

Comment:
Q: I believe we need to make sure that the oracle is the right one, else you can cheat the validity?